engineering-plans/0002-ethereum-tracing-cache.md

```diff
@@ -0,0 +1,184 @@
+
+ There is one flow of data which upholds these invariants. In doing so it makes a tradeoff of increasing latency for the execution of a specific subgraph, but increases throughput of the whole system.
+
+ Within this graph:
+ * Edges which are labelled refer to some subset of the output data.
+ * Edges which are not labelled refer to the entire set of the output data.
+ * Each node executes once for each contiguous range it. That is, it merges all incoming data before executing, and executes the minimum possible times.
+ * The example given is just for 2 addresses. The actual code must work on sets of addresses.
+
+ ```mermaid
+ graph LR;
+    A[Block Range for Contract A & B]
+    A --> |Above Reorg Threshold| E
+    D[Get Cache A]
+    A --> |Below Reorg Threshold A| D
+    A --> |Below Reorg Threshold B| H
+    E[Ethereum A & B]
+    F[Ethereum A]
+    G[Ethereum B]
+    H[Get Cache B]
+    D --> |Found| M
+    H --> |Found| M
+    M[Result]
+    D --> |Missing| N
+    H --> |Missing| N
+    N[Overlap]
+    N --> |A & B| E
+    N --> |A| F
+    N --> |B| G
+    E --> M
+    K[Set Cache A]
+    L[Set Cache B]
+    E --> |B Below Reorg Threshold| L
+    E --> |A Below Reorg Threshold| K
+    F --> K
+    G --> L
+    F --> M
+    G --> M
+ ```
+
+ 
+This construction is designed to make the fewest number of the most efficient calls possible. It is not as complicated as it looks. The actual construction can be expressed as sequential steps with a set of filters preceding eah step.
+
+### Useful dependencies
+The feature deals a lot with ranges and sets. Operations like sum, subtract, merge, and find overlapping are used frequently. [nested_intervals](https://crates.io/crates/nested_intervals) is a crate which provides some of these operations.
+
+
+## Tests
+
+### Benchmark
+A temporary benchmark will be added for indexing a simple subgraph which uses call handlers. The benchmark will be run in these scenarios:
+* Sync before changes
+* Re-sync before changes 
+* Sync after changes
+* Re-sync after changes
+
+### Ranges
+Due to the complexity of the resource minimizing data workflow, it will be useful to have mocks for the cache and database which record their calls, and check that expected calls are made for tricky data sets.
+
+### Database
+A real database integration test will be added to test the add/remove from cache implementation to verify that it correctly merges blocks, handles concurrency issues, etc.
+
+
+## Migration
+
+None
+
+## Documentation
+
+None, aside from code comments
+
+**Estimates:**
+
+These estimates inflated to account for the author's lack of experience with Postgres, Ethereum, Futures0.1, and The Graph in general.
+
+- (1) Create benchmarks
+- Postgres Cache
+  - (0.5) Block Cache
+  - (0.5) Trace Serialization/Deserialization
+  - (1.0) Ranges Cache
+  - (0.5) Concurrency/Transactions
+  - (0.5) Integration tests
+- Data Flow
+  - (3) Implementation
+  - (1) Unit tests
+- (0.5) Run Benchmarks
+
+Total: 8
```

Comment:
We may be underestimating the effort required for integration tests. Since we don't currently have tests that involve indexing subgraphs, this may not be trivial. But perhaps you already had an approach in mind for these?

Reply:
The only 'integration' tests I had in mind were just for using a real database rather than a mock to test the various transactions.

Having a full graph indexing test is possible, but would probably be slow. I see you did some work on this recently.

Reply:
Ah, ok, that makes it more realistic. It'd be fun to have a `graph test` integration test that indexes the same subgraph twice, once without a populated cache and then with a populated cache. We can work that out separately though. 😄